<a href="https://colab.research.google.com/github/nathanbollig/vet-graduate-expectations-survey/blob/main/explore.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Veterinary graduate expectations survey

Exploratory analysis of this data. Start by uploading the data into the working directory. Two files are required:

1.   `SVM.xlsx`: SVM graduate expectations survey results
2.   `WVMA.xlsx`: WVMA graduate expectations survey results

In [ ]:
! pip install xlsxwriter

In [276]:
import pandas as pd
import numpy as np
from scipy.stats import kruskal

## Read in SVM data

In [277]:
# Use top row as header and skip second header row
svm = pd.read_excel('SVM.xlsx', header=0, skiprows=lambda x: x in [1])  

In [278]:
#svm.head(3)

In [279]:
# Read in questions from second header row and associate with column names
question_svm = {}

top_rows = pd.read_excel('SVM.xlsx', nrows=2) 

for col in list(top_rows.columns):
    question_svm[col] = top_rows.iloc[0][col]

## Read in WVMA data

In [280]:
# Use top row as header and skip second header row
wvma = pd.read_excel('WVMA.xlsx', header=0, skiprows=lambda x: x in [1])  

# Read in questions from second header row and associate with column names
question_wvma = {}

top_rows_wvma = pd.read_excel('WVMA.xlsx', nrows=2) 

for col in list(top_rows_wvma.columns):
    question_wvma[col] = top_rows_wvma.iloc[0][col]

## Linking questions

In [281]:
len(question_wvma)

425

In [282]:
len(question_svm)

470

In [283]:
q_svm_set = set(question_svm)
q_wvma_set = set(question_wvma)
len(q_svm_set.intersection(q_wvma_set))

421

There are 425 WVMA questions and 470 SVM questions, and they have 421 questions in common.

In the code below, we loop through all question keys in the SVM results and print results for two potential issues:

*   If the same question key appears in both surveys, but the text of the question differs, there is a "TEXT MISMATCH" message and the two questions are printed.
*   If a question in SVM does not appear in WVMA, the SVM question is printed.



In [284]:
# Review question keys
for svm_key in question_svm.keys():
    if svm_key in question_wvma.keys():
        if question_svm[svm_key] != question_wvma[svm_key]:
            print("\nTEXT MISMATCH! %s" % (svm_key,))
            print("SVM: %s" % (question_svm[svm_key],))
            print("WVMA: %s" % (question_wvma[svm_key],))
    else:
        print("\nSVM key '%s' not in WVMA: %s" % (svm_key, question_svm[svm_key]))


TEXT MISMATCH! Q34
SVM: Select your primary clinical service area. - Selected Choice
WVMA: Select the best description for your practice setting. - Selected Choice

TEXT MISMATCH! Q34_12_TEXT
SVM: Select your primary clinical service area. - Other - Text
WVMA: Select the best description for your practice setting. - Other (please describe) - Text

SVM key 'Q59' not in WVMA: Indicate any AVMA-Recognized Veterinary Specialites in which you currently hold Diplomate status. - Selected Choice

SVM key 'Q59_24_TEXT' not in WVMA: Indicate any AVMA-Recognized Veterinary Specialites in which you currently hold Diplomate status. - Other non-AVMA recognized specialty credentials - Text

TEXT MISMATCH! Q1
SVM: This survey has separate sections detailing expectations for graduates in four species areas.  In each area, rate the level at which you would expect a new graduate to be able to perform the specific procedure or task.  Levels range from "Perform Independently" to "No Expectation to Perform

Now it remains to print WVMA questions that do not appear in the SVM survey. As we saw above, there should be 4 of them.

In [285]:
# Review question keys
for wvma_key in question_wvma.keys():
    if wvma_key not in question_svm.keys():
        print("\nWVMA key '%s' not in SVM: %s" % (wvma_key, question_wvma[wvma_key]))


WVMA key 'Q35' not in SVM: Select the number of veterinarians employed in your practice location.

WVMA key 'Q49' not in SVM: Indicate any AVMA-Recognized Veterinary Specialites in which you currently hold Diplomate status. - Selected Choice

WVMA key 'Q49_24_TEXT' not in SVM: Indicate any AVMA-Recognized Veterinary Specialites in which you currently hold Diplomate status. - Other non-AVMA recognized specialty credentials - Text

WVMA key 'Q46' not in SVM: List any business/practice management, communication, or ethical and professional practices not included in the above lists that you believe are essential for graduates to be able to perform at an independent or indirect supervision level.


In [ ]:
# Show matches
for svm_key in question_svm.keys():
    if svm_key in question_wvma.keys():
        print(svm_key)
            

I examined these lists to develop the following summary. The question keys refer to column names in the first row of the raw spreadsheets (keys in `question_svm` and `question_wvma`, as well as column names in `svm` and `wvma` dataframes).

Comments on mismatched categories in SVM and WVMA

 * `Q34`: Primary clinical service area/practice setting (category list with different categories for SVM vs. WVMA)
 * `Q34_12_TEXT`: Free-text primary clinical service area/practice setting
 * `Q1`: Analogous questions with slightly different wording.
 * `Q43_1` - `Q43_20`: Identical questions (difference due to one space missing before hyphen)

SVM questions with WVMA analogs

 * `svm.Q59`: AVMA-recognized specialty, analogous to `wvma.Q49`
 * `svm.Q59_24_TEXT`: Free-text AVMA-recognized specialty, analogous to `wvma.Q49_24_TEXT`
 * `svm.Q57`: Professional skills, analogous to `wvma.Q46`

SVM-only
 * `svm.Q51_1` - `svm.Q51_22`: DMS rotation length assessment
 * `svm.Q53_1` - `svm.Q53_3`: PBS rotation length assessment
 * `svm.Q54_1` - `svm.Q54_19`: DSS rotation length assessment
 * `svm.Q55`: Free-text which rotations should be extended
 * `svm.Q56`: Free-text which rotations should be shortened

WVMA-only
 * `wvma.Q35`: Number of veterinarians at practice location

### Summary Table

| Question Key             | Linked question key | Description                                                                                               |
| ------------------------ | ------------------- | --------------------------------------------------------------------------------------------------------- |
| Status                   |                     |                                                                                                           |
| IPAddress                |                     |                                                                                                           |
| Progress                 |                     |                                                                                                           |
| Duration (in seconds)    |                     |                                                                                                           |
| Finished                 |                     |                                                                                                           |
| RecordedDate             |                     |                                                                                                           |
| ResponseId               |                     |                                                                                                           |
| RecipientLastName        |                     |                                                                                                           |
| RecipientFirstName       |                     |                                                                                                           |
| RecipientEmail           |                     |                                                                                                           |
| ExternalReference        |                     |                                                                                                           |
| LocationLatitude         |                     |                                                                                                           |
| LocationLongitude        |                     |                                                                                                           |
| DistributionChannel      |                     |                                                                                                           |
| UserLanguage             |                     |                                                                                                           |
| Q34                      |                     | Primary clinical service area/practice setting (category list with different categories for SVM vs. WVMA) |
| Q34\_12\_TEXT            |                     | Free-text primary clinical service area/practice setting                                                  |
| svm.Q59                  | wvma.Q49            | AVMA-recognized specialty                                                                                 |
| svm.Q59\_24\_TEXT        | wvma.Q49\_24\_TEXT  | Free-text AVMA-recognized specialty                                                                       |
| svm.Q57                  | wvma.Q46            | Professional skills                                                                                       |
| Q1                       |                     | Species area                                                                                              |
| Q16\_1 - Q16\_25         |                     | Companion Animal Medical Procedures                                                                       |
| Q17\_1 - Q17\_10         |                     | Companinon Animal Preventive Medicine/Population Health Procedures                                        |
| Q7\_1 - Q7\_25           |                     | Companinon Animal Surgical Procedures                                                                     |
| Q8\_1 - Q8\_8            |                     | Companion Animal Anesthetic Procedures                                                                    |
| Q9\_1 - Q9\_4            |                     | Companion Animal Reproductive Procedures                                                                  |
| Q10\_1 - Q10\_12         |                     | Companion Animal Diagnostic Imaging Procedures                                                            |
| Q11\_1 - Q11\_13         |                     | Companion Animal Clinical Pathology Procedures                                                            |
| Q12\_1 - Q12\_3          |                     | Companion Animal Diagnostic Necropsy Procedures                                                           |
| Q31                      |                     | Free text additional companion animal procedures                                                          |
| Q43\_1 - Q43\_20         |                     | Special Species Medical Procedures                                                                        |
| Q44\_1 - Q44\_9          |                     | Special Species Preventive Medicine/Population Health Procedures                                          |
| Q45\_1 - Q45\_11         |                     | Special Species Surgical Procedures                                                                       |
| Q46\_1 - Q46\_8          |                     | Special Species Anesthetic Procedures                                                                     |
| Q48\_1 - Q48\_6          |                     | Special Species Diagnostic Imaging Procedures                                                             |
| Q49\_1 - Q49\_13         |                     | Special Species Clinical Pathology Procedures                                                             |
| Q50\_1 - Q50\_3          |                     | Special Species Diagnostic Necropsy Procedures                                                            |
| Q51                      |                     | Free text additional special species procedures                                                           |
| Q20\_1 - Q20\_8          |                     | Food Animal Handling and Husbandry Procedures                                                             |
| Q18\_1 - Q18\_27         |                     | Food Animal Medical Procedures                                                                            |
| Q25\_1 - Q25\_16         |                     | Food Animal Surgical Procedures                                                                           |
| Q24\_1 - Q24\_10         |                     | Food Animal Anesthetic Procedures                                                                         |
| Q21\_1 - Q21\_20         |                     | Food Animal Preventive Medicine/Population Health Procedures                                              |
| Q19\_1 - Q19\_11         |                     | Food Animal Reproductive Procedures                                                                       |
| Q23\_1 - Q23\_12         |                     | Food Animal Clinical Pathology Procedures                                                                 |
| Q22\_1 - Q22\_3          |                     | Food Animal Diagnostic Necropsy Procedures                                                                |
| Q27\_1 - Q27\_5          |                     | Food Animal Diagnostic Imaging Procedures                                                                 |
| Q32                      |                     | Free text additional food animal procedures                                                               |
| Q28\_1 - Q28\_7          |                     | Equine Handling and Husbandry Procedures                                                                  |
| Q29\_1 - Q29\_24         |                     | Equine Medical Procedures                                                                                 |
| Q30\_1 - Q30\_8          |                     | Equine Surgical Procedures                                                                                |
| Q31\_1 - Q31\_8          |                     | Equine Anesthetic Procedures                                                                              |
| Q32\_1 - Q32\_15         |                     | Equine Preventive Medicine/Population Health Procedures                                                   |
| Q33\_1 - Q33\_9          |                     | Equine Reproductive Procedures                                                                            |
| Q34\_1 - Q34\_11         |                     | Equine Clinical Pathology Procedures                                                                      |
| Q35\_1 - Q35\_3          |                     | Equine Diagnostic Necropsy Procedures                                                                     |
| Q36\_1 - Q36\_5          |                     | Equine Diagnostic Imaging Procedures                                                                      |
| Q33                      |                     | Free text additional equine procedures                                                                    |
| Q14\_1 - Q14\_6          |                     | Professional and Business Practices                                                                       |
| Q13\_1 - Q13\_11         |                     | Communication Practices                                                                                   |
| Q15\_1 - Q15\_8          |                     | Ethical and Professional Practices                                                                        |
| svm.Q51\_1 - svm.Q51\_22 | none                | DMS rotation length assessment                                                                            |
| svm.Q53\_1 - svm.Q53\_3  | none                | PBS rotation length assessment                                                                            |
| svm.Q54\_1 - svm.Q54\_19 | none                | DSS rotation length assessment                                                                            |
| svm.Q55                  | none                | Free-text which rotations should be extended                                                              |
| svm.Q56                  | none                | Free-text which rotations should be shortened                                                             |
| wvma.Q35                 | none                | Number of veterinarians at practice location                                                              |

## Descriptive Analysis

### Counts of species area

Let's look at the counts of species area (`Q1`) in each survey. First, note that this question allowed multiple responses, which appear as a common-delimited list. The below code counts how many times each species appears, taking into account the possible of multiple responses.

In [287]:
from collections import defaultdict

svm_counts = defaultdict(int) # start each count at zero by default

for entry in list(svm.Q1):
    if isinstance(entry, str):
        species_list = entry.split(',')
        for species in species_list:
            svm_counts[species] += 1
    elif np.isnan(entry) == True:
        svm_counts["empty"] += 1

print("*** SVM Survey ***")
for key, val in svm_counts.items():
    print("%s: %i" % (key, val))

*** SVM Survey ***
Companion Animal (canine and/or feline): 46
Food Animal (bovine): 18
Equine: 16
Special Species: 19
empty: 5


In [288]:
wvma_counts = defaultdict(int) # start each count at zero by default

for entry in list(wvma.Q1):
    if isinstance(entry, str):
        species_list = entry.split(',')
        for species in species_list:
            wvma_counts[species] += 1
    elif np.isnan(entry) == True:
        wvma_counts["empty"] += 1

print("*** WVMA Survey ***")
for key, val in wvma_counts.items():
    print("%s: %i" % (key, val))

*** WVMA Survey ***
Companion Animal (canine and/or feline): 115
Food Animal (bovine): 48
Equine: 30
Special Species (ex. exotic companion animals): 21
empty: 29


### Note about organization

There are several levels of organization in our interpretation of this data.

 * `Group`: One of the 4 species groups (companion animal, special species, food animal, or equine)
     * `Question`: A group of procedures in a category such as "Medical Procedures" or "Surgical Procedures"
          * `Sub-question`: A particular procedure

We can perform analysis at the sub-question level, or pool upwards to the question or group level. I will do all of this below.





### Examine sub-questions within companion animal Q16

Now let's consider Companion Animal Medical Procedures (`Q16_1` - `Q16_25`). Below is code to convert these responses to ordinal representations and compare the distributions between SVM and WVMA.

In [289]:
svm_strings = []
wvma_strings = []

for i in range(1,26):
    qkey = "Q16_" + str(i)
    svm_strings.extend(list(svm[qkey]))
    wvma_strings.extend(list(wvma[qkey]))

In [290]:
from collections import Counter

Counter(svm_strings)

Counter({'No Expectation to Perform Procedure': 51,
         'Perform Independently': 740,
         'Perform with Assistance (assist with portions of procedure)': 18,
         'Perform with Direct Supervision (present in room during procedure)': 50,
         'Perform with Indirect Supervision (available in building or by phone if needed)': 140,
         nan: 501})

In [291]:
Counter(wvma_strings)

Counter({'No Expectation to Perform Procedure': 53,
         'Perform Independently': 1207,
         'Perform with Assistance (assist with portions of procedure)': 146,
         'Perform with Direct Supervision (present in room during procedure)': 344,
         'Perform with Indirect Supervision (available in building or by phone if needed)': 707,
         nan: 1918})

Let's encode the expectation response in the following way:

 * 0: No Expectation to Perform Procedure

 * 1: Perform with Assistance (assist with portions of procedure)
 
 * 2: Perform with Direct Supervision (present in room during procedure)

 * 3: Perform with Indirect Supervision (available in building or by phone if needed)

 * 4: Perform Independently

In [292]:
def encode_expectation(response_string):
    if isinstance(response_string, int) == True:
        return response_string
    
    # Encode nan values as -1
    if isinstance(response_string, str) == False:
        if np.isnan(response_string) == True:
            return -1
    
    # Encode string
    s = response_string.lower()
    if s.find('no expectation') > -1:
        return 0
    elif s.find('with assistance') > -1:
        return 1
    elif s.find('indirect supervision') > -1:
        return 3
    elif s.find('direct supervision') > -1:
        return 2
    elif s.find('independently') > -1:
        return 4
    else:
        print(response_string)
        raise ValueError('Expected performance response was not formatted as expected.')

In [293]:
# Filter dataframe to only companion animal respondants (may have responded to other species too)
ca_svm = svm[svm['Q1'].str.contains('Companion Animal (canine and/or feline)', na=False, regex=False)].copy()

In [294]:
# Filter dataframe to only companion animal respondants (may have responded to other species too)
ca_wvma = wvma[wvma['Q1'].str.contains('Companion Animal (canine and/or feline)', na=False, regex=False)].copy()

Below we see that of the 46 companion animal responses for svm.Q16_1, 6 were blank.

In [295]:
ca_svm.Q16_1.value_counts(dropna=False)

Perform Independently                  39
NaN                                     6
No Expectation to Perform Procedure     1
Name: Q16_1, dtype: int64

Let's summarize counts of responses.

In [296]:
# Tally data for this question group
NUM_QUESTIONS = 25
svm_counts = np.zeros((NUM_QUESTIONS, 6), dtype=int) # Row for each question, column for empty (-1), 0, 1, 2, 3, and 4 responses
wvma_counts = np.zeros((NUM_QUESTIONS, 6), dtype=int) # Row for each question, column for empty (-1), 0, 1, 2, 3, and 4 responses
rows = []

for i in range(1,26):
    qkey = "Q16_" + str(i)
    qstring = question_svm[qkey].split('-')[2] # could refer to questions_svm or questions_wvma

    # Encoding
    ca_svm[qkey] = ca_svm[qkey].apply(lambda x: encode_expectation(x))
    ca_wvma[qkey] = ca_wvma[qkey].apply(lambda x: encode_expectation(x))

    # SVM tally
    counts = ca_svm[qkey].value_counts(dropna=False)
    for key in counts.keys():
        svm_counts[i-1][key+1] += counts[key] # question index is 1-based; keys range from -1 to 4
    counts = svm_counts[i-1][1:] # counts of 0, 1, 2, 3, and 4
    svm_mean = (0*counts[0] + 1*counts[1] + 2*counts[2] + 3*counts[3] + 4*counts[4]) / np.sum(counts)

    # WVMA tally
    counts = ca_wvma[qkey].value_counts(dropna=False)
    for key in counts.keys():
        wvma_counts[i-1][key+1] += counts[key]
    counts = wvma_counts[i-1][1:] # counts of 0, 1, 2, 3, and 4
    wvma_mean = (0*counts[0] + 1*counts[1] + 2*counts[2] + 3*counts[3] + 4*counts[4]) / np.sum(counts)
      
    # Cache for table of results
    row = [qstring] + list(svm_counts[i-1]) + [svm_mean] + list(wvma_counts[i-1]) + [wvma_mean]
    rows.append(row)

In [297]:
# Assemble table of results
q16_table = pd.DataFrame(rows, columns=["Question", "SVM: empty", "SVM: 0", "SVM: 1", "SVM: 2", "SVM: 3", "SVM: 4", "SVM: avg", "WVMA: empty", "WVMA: 0", "WVMA: 1", "WVMA: 2", "WVMA: 3", "WVMA: 4", "WVMA: avg"])
q16_table

Question  ...  WVMA: avg
0              Obtain history and perform complete PE  ...   3.878788
1                             Perform ophthalmic exam  ...   3.240000
2                              Perform otoscopic exam  ...   3.670000
3                             Perform neurologic exam  ...   3.040000
4                             Perform orthopedic exam  ...   3.050000
5    Develop problem list and rank order different...  ...   3.400000
6               Develop and interpret diagnostic plan  ...   2.989899
7                              Develop treatment plan  ...   2.919192
8                         Calculate medication dosage  ...   3.696970
9                                  Write prescription  ...   3.626263
10                               Calculate fluid rate  ...   3.494949
11   Calculate constant rate infusion medication r...  ...   2.948980
12                         Perform IV catheterization  ...   3.193878
13                               Perform cystocenesis  ...   2.989691
14           Perform thoracocentesis/abdominocentesis  ...   1.979592
15                               Perform venipuncture  ...   3.597938
16                            Insert nasogastric tube  ...   1.918367
17                      Perform anal gland expression  ...   3.690722
18                                Bandage application  ...   3.234694
19                                 Perform euthanasia  ...   3.350515
20                 Auscultate and grade heart murmurs  ...   3.185567
21   Recognize zoonotic diseases and respond accor...  ...   2.989691
22   Triages cases to address most urgent and impo...  ...   3.082474
23   Recognizes emergent situation and directs action  ...   3.051546
24   Identifies situations in which referral is wa...  ...   2.948454

[25 rows x 15 columns]

## Companion Animal Medical Q16 Subquestions

Let's apply Kruskal-Wallis H-test to each subquestion in the Q16 question.

In [298]:
# Tally data for this question group
NUM_QUESTIONS = 25
ALPHA = 0.05
svm_counts = np.zeros((NUM_QUESTIONS, 6), dtype=int) # Row for each question, column for empty (-1), 0, 1, 2, 3, and 4 responses
wvma_counts = np.zeros((NUM_QUESTIONS, 6), dtype=int) # Row for each question, column for empty (-1), 0, 1, 2, 3, and 4 responses
rows = []
svm_pooled = []
wvma_pooled = []

for i in range(1,26):
    qkey = "Q16_" + str(i)
    qstring = question_svm[qkey].split('-')[2] # could refer to questions_svm or questions_wvma

    # Encoding
    ca_svm[qkey] = ca_svm[qkey].apply(lambda x: encode_expectation(x))
    ca_wvma[qkey] = ca_wvma[qkey].apply(lambda x: encode_expectation(x))

    # SVM tally
    counts = ca_svm[qkey].value_counts(dropna=False)
    for key in counts.keys():
        svm_counts[i-1][key+1] += counts[key] # question index is 1-based; keys range from -1 to 4
    counts = svm_counts[i-1][1:] # counts of 0, 1, 2, 3, and 4
    svm_mean = (0*counts[0] + 1*counts[1] + 2*counts[2] + 3*counts[3] + 4*counts[4]) / np.sum(counts)

    # WVMA tally
    counts = ca_wvma[qkey].value_counts(dropna=False)
    for key in counts.keys():
        wvma_counts[i-1][key+1] += counts[key]
    counts = wvma_counts[i-1][1:] # counts of 0, 1, 2, 3, and 4
    wvma_mean = (0*counts[0] + 1*counts[1] + 2*counts[2] + 3*counts[3] + 4*counts[4]) / np.sum(counts)
    
    # Get data
    svm_data = list(ca_svm[qkey])
    wvma_data = list(ca_wvma[qkey])

    # Remove empty values from data
    svm_data = [x for x in svm_data if x != -1]
    wvma_data = [x for x in wvma_data if x != -1]

    # compare samples
    stat, p = kruskal(svm_data, wvma_data)

    # Determine significance
    if p > ALPHA:
        sig = ""
    else:
        sig = "*"

    # Cache for pooled data
    svm_pooled.extend(svm_data)
    wvma_pooled.extend(wvma_data)

    # Cache for table of results
    row = [qstring] + list(svm_counts[i-1]) + [svm_mean] + list(wvma_counts[i-1]) + [wvma_mean, stat, p, sig]
    rows.append(row)

In [299]:
# Assemble table of results
q16_table = pd.DataFrame(rows, columns=["Subquestion", "SVM: empty", "SVM: 0", "SVM: 1", "SVM: 2", "SVM: 3", "SVM: 4", "SVM: avg", "WVMA: empty", "WVMA: 0", "WVMA: 1", "WVMA: 2", "WVMA: 3", " WVMA: 4", "WVMA: avg", "stat", "pval", "sig"])
q16_table

Subquestion  ...  sig
0              Obtain history and perform complete PE  ...     
1                             Perform ophthalmic exam  ...     
2                              Perform otoscopic exam  ...     
3                             Perform neurologic exam  ...    *
4                             Perform orthopedic exam  ...    *
5    Develop problem list and rank order different...  ...    *
6               Develop and interpret diagnostic plan  ...    *
7                              Develop treatment plan  ...    *
8                         Calculate medication dosage  ...     
9                                  Write prescription  ...    *
10                               Calculate fluid rate  ...     
11   Calculate constant rate infusion medication r...  ...    *
12                         Perform IV catheterization  ...     
13                               Perform cystocenesis  ...    *
14           Perform thoracocentesis/abdominocentesis  ...    *
15                               Perform venipuncture  ...     
16                            Insert nasogastric tube  ...    *
17                      Perform anal gland expression  ...     
18                                Bandage application  ...     
19                                 Perform euthanasia  ...     
20                 Auscultate and grade heart murmurs  ...    *
21   Recognize zoonotic diseases and respond accor...  ...    *
22   Triages cases to address most urgent and impo...  ...    *
23   Recognizes emergent situation and directs action  ...    *
24   Identifies situations in which referral is wa...  ...    *

[25 rows x 18 columns]

These results indicate that 15/25 procedures in these groups have significantly different expectations between SVM and WVMA. In all cases, the SVM expectations reflect a higher level of independent ability than WVMA expectations.

In [300]:
stat, p = kruskal(svm_pooled, wvma_pooled)
diff_mean = np.mean(svm_pooled) - np.mean(wvma_pooled)
print('Pooled data: stat=%.3f, p=%.2e, diff_mean (svm-wvma)=%.3f' % (stat, p, diff_mean))

Pooled data: stat=149.813, p=1.90e-34, diff_mean (svm-wvma)=0.334


When responses are pooled across the set of subquestions in this group, we see that SVM respondants indicated a higher degree of independence compared to WVMA respondants overall.

## Reusable code for question analysis

We will refactor the above into reusable code.

In [301]:
def analyze_question(question_number, filtered_svm_df, filtered_wvma_df, n_subquestions, alpha=0.05, verbose=True):
    """
    Perform an analysis of a given question on a species-filtered dataframe.
    
    Inputs:
        question_number: main question number to analyze
        filtered_svm_df: SVM dataframe filtered to respondants with the desired species area
        filtered_wvma_df: WVMA dataframe filtered to respondants with the desired species area
        n_subquestions: number of subquestions in the main question
        alpha: power level for the statistical test

    Prints a summary of results.

    Outputs:
        table: summary table
        (pooled_stat, pooled_p, pooled_diff_mean): tuple of statistics describing output of Kruskal test on data pooled across subquestions
        svm_data: list of pooled SVM data
        wvma_data: list of pooled WVMA data
        sig_count: number of subquestions with significant difference detected (between SVM and WVMA responses), according to Kruskal test applied at subquestion level

    """

    svm_counts = np.zeros((n_subquestions, 6), dtype=int) # Row for each question, column for empty (-1), 0, 1, 2, 3, and 4 responses
    wvma_counts = np.zeros((n_subquestions, 6), dtype=int) # Row for each question, column for empty (-1), 0, 1, 2, 3, and 4 responses
    rows = []
    svm_pooled = []
    wvma_pooled = []
    sig_count = 0

    for i in range(1, n_subquestions+1):
        qkey = "Q" + str(question_number) + "_" + str(i)
        qstring = question_svm[qkey].split('-')[2] # could refer to questions_svm or questions_wvma

        # Encoding
        filtered_svm_df[qkey] = filtered_svm_df[qkey].apply(lambda x: encode_expectation(x))
        filtered_wvma_df[qkey] = filtered_wvma_df[qkey].apply(lambda x: encode_expectation(x))

        # SVM tally
        counts = filtered_svm_df[qkey].value_counts(dropna=False)
        for key in counts.keys():
            svm_counts[i-1][key+1] += counts[key] # question index is 1-based; keys range from -1 to 4
        counts = svm_counts[i-1][1:] # counts of 0, 1, 2, 3, and 4
        svm_num_responses = np.sum(counts)
        svm_mean = (0*counts[0] + 1*counts[1] + 2*counts[2] + 3*counts[3] + 4*counts[4]) / svm_num_responses

        # WVMA tally
        counts = filtered_wvma_df[qkey].value_counts(dropna=False)
        for key in counts.keys():
            wvma_counts[i-1][key+1] += counts[key]
        counts = wvma_counts[i-1][1:] # counts of 0, 1, 2, 3, and 4
        wvma_num_responses = np.sum(counts)
        wvma_mean = (0*counts[0] + 1*counts[1] + 2*counts[2] + 3*counts[3] + 4*counts[4]) / wvma_num_responses
        
        # Get data
        svm_data = list(filtered_svm_df[qkey])
        wvma_data = list(filtered_wvma_df[qkey])

        # Remove empty values from data
        svm_data = [x for x in svm_data if x != -1]
        wvma_data = [x for x in wvma_data if x != -1]

        assert(svm_num_responses == len(svm_data))
        assert(wvma_num_responses == len(wvma_data))

        # compare samples
        stat, p = kruskal(svm_data, wvma_data)

        # Determine significance
        if p > ALPHA:
            sig = ""
        else:
            sig = "*"
            sig_count += 1

        # Cache for pooled data
        svm_pooled.extend(svm_data)
        wvma_pooled.extend(wvma_data)

        # Cache for table of results
        row = [qstring] + list(svm_counts[i-1]) + [svm_mean, svm_num_responses] + list(wvma_counts[i-1]) + [wvma_mean, wvma_num_responses, svm_mean-wvma_mean, stat, p, sig]
        rows.append(row)

    # Assemble table of results
    table = pd.DataFrame(rows, columns=["Subquestion", "SVM: empty", "SVM: 0", "SVM: 1", "SVM: 2", "SVM: 3", "SVM: 4", "SVM: avg", "SVM: num responses", "WVMA: empty", "WVMA: 0", "WVMA: 1", "WVMA: 2", "WVMA: 3", "WVMA: 4", "WVMA: avg", "WVMA: num responses", "Diff Mean (svm-wvma)", "stat", "pval", "sig"])

    # Apply Kruskal test to pooled data
    pooled_stat, pooled_p = kruskal(svm_pooled, wvma_pooled)
    pooled_diff_mean = np.mean(svm_pooled) - np.mean(wvma_pooled)

    # Print
    if verbose == True:
        print('Pooled Q%s: stat=%.3f, p=%.2e, diff_mean (svm-wvma)=%.3f, sig_subq=%s/%s' % (question_number, pooled_stat, pooled_p, pooled_diff_mean, sig_count, n_subquestions))

    return table, (pooled_stat, pooled_p, pooled_diff_mean), svm_pooled, wvma_pooled, sig_count


The above code prints a question-level summary and returns a table of details for particular sub-questions.

In [302]:
table, subq_pooled_result, svm_data, wvma_data, sig_count = analyze_question(16, ca_svm, ca_wvma, n_subquestions=25)

Pooled Q16: stat=149.813, p=1.90e-34, diff_mean (svm-wvma)=0.334, sig_subq=15/25


In [303]:
table

Subquestion  ...  sig
0              Obtain history and perform complete PE  ...     
1                             Perform ophthalmic exam  ...     
2                              Perform otoscopic exam  ...     
3                             Perform neurologic exam  ...    *
4                             Perform orthopedic exam  ...    *
5    Develop problem list and rank order different...  ...    *
6               Develop and interpret diagnostic plan  ...    *
7                              Develop treatment plan  ...    *
8                         Calculate medication dosage  ...     
9                                  Write prescription  ...    *
10                               Calculate fluid rate  ...     
11   Calculate constant rate infusion medication r...  ...    *
12                         Perform IV catheterization  ...     
13                               Perform cystocenesis  ...    *
14           Perform thoracocentesis/abdominocentesis  ...    *
15                               Perform venipuncture  ...     
16                            Insert nasogastric tube  ...    *
17                      Perform anal gland expression  ...     
18                                Bandage application  ...     
19                                 Perform euthanasia  ...     
20                 Auscultate and grade heart murmurs  ...    *
21   Recognize zoonotic diseases and respond accor...  ...    *
22   Triages cases to address most urgent and impo...  ...    *
23   Recognizes emergent situation and directs action  ...    *
24   Identifies situations in which referral is wa...  ...    *

[25 rows x 21 columns]

## Companion Animal Group

Recall that the companion animal group contains a set of questions. Now let's use the modularized question analysis code to perform a similar analysis across all questions in the companion animal group.

In [304]:
# cache data across all groups
group_data = []
group_columns = ["Group", "Pooled stat", "Pooled p", "Pooled diff_mean (svm-wvma)", "Num questions", "Fraction of sig questions", "Pooled num SVM reponses", "Pooled num WVMA responses"]

In [305]:
# cache tables
output_tables = []
output_tables_sheet_names = []

# cache subquestion table data
output_subq_data = []

In [306]:
# Input info about question group

question_list = [16,17,7,8,9,10,11,12]
n_subq_list = [25,10,25,8,4,12,13,3]
question_strings = ['Medical Procedures',
                    'Preventive Medicine/Population Health Procedures',
                    'Surgical Procedures', 
                    'Anesthetic Procedures', 
                    'Reproductive Procedures',
                    'Diagnostic Imaging Procedures',
                    'Clinical Pathology Procedures',
                    'Diagnostic Necropsy Procedures']

assert(len(question_list) == len(n_subq_list))
assert(len(n_subq_list) == len(question_strings))

In [307]:
# Code to analyze all questions within the group

def analyze_group(question_list, n_subq_list, question_strings, filtered_svm_df, filtered_wvma_df):
    svm_pooled = [] # now pooling over entire group
    wvma_pooled = []
    rows = []
    sig_count = 0
    subq_tables = []
    subq_tables_names = []

    for i in range(len(question_list)):
        question_number = question_list[i]
        n_subquestions = n_subq_list[i]
        question_string = question_strings[i]

        # Run analysis
        table, subq_pooled_result, svm_data, wvma_data, sig_subq = analyze_question(question_number, filtered_svm_df, filtered_wvma_df, n_subquestions, verbose=False)
        pooled_stat, pooled_p, pooled_diff_mean = subq_pooled_result
        svm_num_responses = len(svm_data)
        wvma_num_responses = len(wvma_data)

        # Cache procedure tables
        subq_tables.append(table)
        subq_tables_names.append('Q'+str(question_number))

        # Pool
        svm_pooled.extend(svm_data)
        wvma_pooled.extend(wvma_data)

        # Determine significance
        if pooled_p > ALPHA:
            sig = ""
        else:
            sig = "*"
            sig_count += 1

        # Cache data for group summary
        row = ['Q'+str(question_number), question_string, pooled_stat, pooled_p, sig, pooled_diff_mean, n_subquestions, "%i/%i" % (sig_subq,n_subquestions), svm_num_responses, wvma_num_responses]
        rows.append(row)

    # Assemble table of results
    group_table = pd.DataFrame(rows, columns=["Question number", "Category", "Pooled stat", "Pooled p", "Sig", "Pooled Diff Mean (svm-wvma)", "Num subquestions", "Fraction of sig subquestions", "Pooled num SVM responses", "Pooled num WVMA responses"])                     

    # Apply Kruskal test to pooled data
    pooled_stat, pooled_p = kruskal(svm_pooled, wvma_pooled)
    pooled_diff_mean = np.mean(svm_pooled) - np.mean(wvma_pooled)

    # Print
    print('Group result (all questions): stat=%.3f, p=%.2e, diff_mean (svm-wvma)=%.3f, sig_subq=%s/%s' % (pooled_stat, pooled_p, pooled_diff_mean, sig_count, len(question_list)))

    return group_table, (pooled_stat, pooled_p, pooled_diff_mean), sig_count, len(question_list), len(svm_pooled), len(wvma_pooled), (subq_tables, subq_tables_names)

In [308]:
group_table, pooled_q_stats, sig_count, n_questions, svm_responses, wvma_responses, subq_data  = analyze_group(question_list, n_subq_list, question_strings, ca_svm, ca_wvma)
pooled_stat, pooled_p, pooled_diff_mean = pooled_q_stats
group_data.append(["Companion Animal", pooled_stat, pooled_p, pooled_diff_mean, n_questions, "%i/%i" % (sig_count,n_questions), svm_responses, wvma_responses])
output_tables.append(group_table)
output_tables_sheet_names.append("Companion Animal")
output_subq_data.append(subq_data)
group_table

Group result (all questions): stat=141.880, p=1.03e-32, diff_mean (svm-wvma)=0.235, sig_subq=6/8


Question number  ... Pooled num WVMA responses
0             Q16  ...                      2457
1             Q17  ...                       930
2              Q7  ...                      2316
3              Q8  ...                       719
4              Q9  ...                       356
5             Q10  ...                      1042
6             Q11  ...                      1125
7             Q12  ...                       261

[8 rows x 10 columns]

## Special Species Group

In [309]:
# Filter dataframes to only companion animal respondants (may have responded to other species too)
ss_svm = svm[svm['Q1'].str.contains('Special Species', na=False, regex=False)].copy()
ss_wvma = wvma[wvma['Q1'].str.contains('Special Species', na=False, regex=False)].copy()

In [310]:
# Input info about question group

question_list = [43, 44, 45, 46, 48, 49, 50]
n_subq_list = [20, 9, 11, 8, 6, 13, 3]
question_strings = ['Medical Procedures',
                    'Preventive Medicine/Population Health Procedures',
                    'Surgical Procedures', 
                    'Anesthetic Procedures', 
                    'Diagnostic Imaging Procedures',
                    'Clinical Pathology Procedures',
                    'Diagnostic Necropsy Procedures']

assert(len(question_list) == len(n_subq_list))
assert(len(n_subq_list) == len(question_strings))

In [311]:
group_table, pooled_q_stats, sig_count, n_questions, svm_responses, wvma_responses, subq_data  = analyze_group(question_list, n_subq_list, question_strings, ss_svm, ss_wvma)
pooled_stat, pooled_p, pooled_diff_mean = pooled_q_stats
group_data.append(["Special Species", pooled_stat, pooled_p, pooled_diff_mean, n_questions, "%i/%i" % (sig_count,n_questions), svm_responses, wvma_responses])
output_tables.append(group_table)
output_tables_sheet_names.append("Special Species")
output_subq_data.append(subq_data)
group_table

Group result (all questions): stat=140.399, p=2.18e-32, diff_mean (svm-wvma)=0.742, sig_subq=5/7


Question number  ... Pooled num WVMA responses
0             Q43  ...                       340
1             Q44  ...                       148
2             Q45  ...                       185
3             Q46  ...                       128
4             Q48  ...                        96
5             Q49  ...                       208
6             Q50  ...                        48

[7 rows x 10 columns]

## Food Animal Group

In [312]:
# Filter dataframes to only companion animal respondants (may have responded to other species too)
fa_svm = svm[svm['Q1'].str.contains('Food Animal', na=False, regex=False)].copy()
fa_wvma = wvma[wvma['Q1'].str.contains('Food Animal', na=False, regex=False)].copy()

In [313]:
# Input info about question group

question_list = [20, 18, 25, 24, 21, 19, 23, 22, 27]
n_subq_list = [8, 27, 16, 10, 20, 11, 12, 3, 5]
question_strings = ['Handling and Husbandry Procedures',
                    'Medical Procedures',
                    'Surgical Procedures',
                    'Anesthetic Procedures',
                    'Preventive Medicine/Population Health Procedures',
                    'Reproductive Procedures',
                    'Clinical Pathology Procedures',
                    'Diagnostic Necropsy Procedures',
                    'Diagnostic Imaging Procedures']


assert(len(question_list) == len(n_subq_list))
assert(len(n_subq_list) == len(question_strings))

In [314]:
group_table, pooled_q_stats, sig_count, n_questions, svm_responses, wvma_responses, subq_data  = analyze_group(question_list, n_subq_list, question_strings, fa_svm, fa_wvma)
pooled_stat, pooled_p, pooled_diff_mean = pooled_q_stats
group_data.append(["Food Animal", pooled_stat, pooled_p, pooled_diff_mean, n_questions, "%i/%i" % (sig_count,n_questions), svm_responses, wvma_responses])
output_tables.append(group_table)
output_tables_sheet_names.append("Food Animal")
output_subq_data.append(subq_data)
group_table

Group result (all questions): stat=11.932, p=5.52e-04, diff_mean (svm-wvma)=0.070, sig_subq=2/9


Question number  ... Pooled num WVMA responses
0             Q20  ...                       295
1             Q18  ...                       997
2             Q25  ...                       574
3             Q24  ...                       360
4             Q21  ...                       718
5             Q19  ...                       385
6             Q23  ...                       418
7             Q22  ...                       105
8             Q27  ...                       174

[9 rows x 10 columns]

## Equine Group

In [315]:
# Filter dataframes to only companion animal respondants (may have responded to other species too)
eq_svm = svm[svm['Q1'].str.contains('Equine', na=False, regex=False)].copy()
eq_wvma = wvma[wvma['Q1'].str.contains('Equine', na=False, regex=False)].copy()

In [316]:
# Input info about question group

question_list = [28, 29, 30, 31, 32, 33, 34, 35, 36]
n_subq_list = [7, 24, 8, 8, 15, 9, 11, 3, 5]
question_strings = ['Handling and Husbandry Procedures',
                    'Medical Procedures',
                    'Surgical Procedures',
                    'Anesthetic Procedures',
                    'Preventive Medicine/Population Health Procedures',
                    'Reproductive Procedures',
                    'Clinical Pathology Procedures',
                    'Diagnostic Necropsy Procedures',
                    'Diagnostic Imaging Procedures']


assert(len(question_list) == len(n_subq_list))
assert(len(n_subq_list) == len(question_strings))

In [317]:
group_table, pooled_q_stats, sig_count, n_questions, svm_responses, wvma_responses, subq_data  = analyze_group(question_list, n_subq_list, question_strings, eq_svm, eq_wvma)
pooled_stat, pooled_p, pooled_diff_mean = pooled_q_stats
group_data.append(["Equine", pooled_stat, pooled_p, pooled_diff_mean, n_questions, "%i/%i" % (sig_count,n_questions), svm_responses, wvma_responses])
output_tables.append(group_table)
output_tables_sheet_names.append("Equine")
output_subq_data.append(subq_data)
group_table

Group result (all questions): stat=22.572, p=2.02e-06, diff_mean (svm-wvma)=0.161, sig_subq=3/9


Question number  ... Pooled num WVMA responses
0             Q28  ...                       175
1             Q29  ...                       599
2             Q30  ...                       200
3             Q31  ...                       200
4             Q32  ...                       360
5             Q33  ...                       216
6             Q34  ...                       253
7             Q35  ...                        68
8             Q36  ...                       111

[9 rows x 10 columns]

## Group Summary

In [318]:
group_summary_table = pd.DataFrame(group_data, columns=group_columns)

In [319]:
pvals = list(group_summary_table['Pooled p'])

sigs = []
for p in pvals:
  if p > ALPHA:
      sig = ""
  else:
      sig = "*"
  sigs.append(sig)

group_summary_table.insert(loc=3, column='Sig', value=sigs)

In [320]:
# Add group summary to the beginning of output tables
output_tables.insert(0, group_summary_table)
output_tables_sheet_names.insert(0, "Group summary")

In [321]:
group_summary_table

Group  ...  Pooled num WVMA responses
0  Companion Animal  ...                       9206
1   Special Species  ...                       1153
2       Food Animal  ...                       4026
3            Equine  ...                       2182

[4 rows x 9 columns]

Looking at pooled comparisons at the group level, there are significant differences between the SVM and WVMA expectations for all species groups.

As we saw, there are significant differences at the individual question and subquestion levels within all groups as well.

# Generate tables

We will generate the following tables using pooled data from these experiments:

1.   `summary.xlsx`: Group summary table and a table for procedure sets (questions) within each group.
2.   `companion_animal.xlsx`: Tables for all procedures within the companion animal group.
3.   `special_species.xlsx`: Tables for all procedures within the special species group.
4.   `food_animal.xlsx`: Tables for all procedures within the food animal group.
5.   `equine.xlsx`:Tables for all procedures within the equine group.



## Summary

In [322]:
writer = pd.ExcelWriter('summary.xlsx', engine='xlsxwriter')

for i,table in enumerate(output_tables):
    sheet_name = output_tables_sheet_names[i]
    table.to_excel(writer, sheet_name=sheet_name, index=False)

    # Auto-adjust columns widths
    for column in table:
        column_width = max(table[column].astype(str).map(len).max(), len(column))
        col_idx = table.columns.get_loc(column)
        writer.sheets[sheet_name].set_column(col_idx, col_idx, column_width)

writer.save()

## All procedures

In [323]:
for i, file in enumerate(['companion_animal.xlsx', 'special_species.xlsx', 'food_animal.xlsx', 'equine.xlsx']):
    subq_data = output_subq_data[i]
    subq_tables, subq_tables_names = subq_data

    # Loop through tables
    writer = pd.ExcelWriter(file, engine='xlsxwriter')

    for i,table in enumerate(subq_tables):
        sheet_name = subq_tables_names[i]
        table.to_excel(writer, sheet_name=sheet_name, index=False)

        # Auto-adjust columns widths
        for column in table:
            column_width = max(table[column].astype(str).map(len).max(), len(column))
            col_idx = table.columns.get_loc(column)
            writer.sheets[sheet_name].set_column(col_idx, col_idx, column_width)

    writer.save()